page_iterator.go

```diff
@@ -0,0 +1,77 @@
+package msgraphgocore
+
+import "net/http"
+
+type Item interface {
+	GetDisplayName() string
+}
+
+type Page interface {
+	GetValue() []Item
+	GetNextLink() string
+	GetNextPage() Page
+}
+
+type PageIterator struct {
+	page       Page
+	client     http.Client
+	pauseIndex int
+}
+
+func NewPageIterator(page Page, client http.Client) *PageIterator {
+	//TODO: Pass client to page
+
+	return &PageIterator{
+		page,
+		client,
+		0, // pauseIndex helps us remember where we paused enumeration in the page.
+	}
+}
+
+func (pI *PageIterator) HasNext() bool {
+	if pI.page.GetNextLink() == "" {
```

Comment:
should check whether page or pI are nil too

Reply:
it might be a misunderstanding from my side (fairly new to Go) but the method can technically be called on a nil pointer can't it?

Reply:
Yes it can which will cause a panic during build. I think that's behaviour is fine because we are not returning default values.

Reply:
interesting, I didn't know it'd panic the build for the consumer if they did that.
I've also seen this style for functions `func (pI PageIterator) HasNext() bool` (note the lack of star symbol). Is there any difference then if the compiler guards against calling the "pointer ones" being called with nil?

Reply:
I don't understand your question.

Reply:
is there any difference in behavior between
`func (pI PageIterator) HasNext() bool`
and
`func (pI *PageIterator) HasNext() bool`

since you mentioned that calling `func (pI *PageIterator) HasNext() bool` with `nil` is not possible due to compile time validation.